yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,1048 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history.yarn
+
+import java.net.{ConnectException, URI}
+import java.util.concurrent.atomic.{AtomicLong, AtomicBoolean, AtomicInteger}
+import java.util.concurrent.{TimeUnit, LinkedBlockingDeque}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import org.apache.hadoop.security.UserGroupInformation
+import org.apache.hadoop.yarn.api.records.timeline.{TimelineDomain, TimelineEntity, TimelineEvent, TimelinePutResponse}
+import org.apache.hadoop.yarn.api.records.{ApplicationAttemptId, ApplicationId}
+import org.apache.hadoop.yarn.client.api.TimelineClient
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+
+import org.apache.spark.deploy.history.yarn.YarnTimelineUtils._
+import org.apache.spark.scheduler._
+import org.apache.spark.scheduler.cluster.{YarnExtensionService, YarnExtensionServiceBinding}
+import org.apache.spark.util.{SystemClock, Utils}
+import org.apache.spark.{Logging, SparkContext}
+
+/**
+ * A Yarn Extension Service to post lifecycle events to a registered
+ * YARN Timeline Server.
+ */
+private[spark] class YarnHistoryService extends YarnExtensionService with Logging {
+
+  import org.apache.spark.deploy.history.yarn.YarnHistoryService._
+
+  /** Simple state model implemented in an atomic integer */
+  private val _serviceState = new AtomicInteger(CreatedState)
+
+  def serviceState: Int = {
+    _serviceState.get()
+  }
+  def enterState(state: Int): Int = {
+    logDebug(s"Entering state $state from $serviceState")
+    _serviceState.getAndSet(state)
+  }
+
+  /**
+   * Spark context; valid once started
+   */
+  private var sparkContext: SparkContext = _
+
+  /** YARN configuration from the spark context */
+  private var config: YarnConfiguration = _
+
+  /** application ID. */
+  private var _applicationId: ApplicationId = _
+
+  /** attempt ID this will be null if the service is started in yarn-client mode */
+  private var _attemptId: Option[ApplicationAttemptId] = None
+
+  /** YARN timeline client */
+  private var _timelineClient: Option[TimelineClient] = None
+
+  /** registered event listener */
+  private var listener: Option[YarnEventListener] = None
+
+  /** Application name  from the spark start event */
+  private var applicationName: String = _
+
+  /** Application ID*/
+  private var sparkApplicationId: Option[String] = None
+
+  /** Optional Attempt ID from the spark start event */
+  private var sparkApplicationAttemptId: Option[String] = None
+
+  /** user name as derived from `SPARK_USER` env var or `UGI` */
+  private var userName = Utils.getCurrentUserName()
+
+  /** Clock for recording time */
+  private val clock = new SystemClock()
+
+  /**
```

Comment:
nit: please be consistent re: single line vs. multi-line docs, and starting with capital or lower case letters.